// react-pdf-highlighter-plus/dist/esm/export-pdf-W2QGWADM.js
// src/lib/export-pdf.ts
import { PDFDocument, rgb, StandardFonts } from "pdf-lib";
function parseColor(color) {
  const rgbaMatch = color.match(
    /rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*([\d.]+))?\)/
  );
  if (rgbaMatch) {
    return {
      r: parseInt(rgbaMatch[1]) / 255,
      g: parseInt(rgbaMatch[2]) / 255,
      b: parseInt(rgbaMatch[3]) / 255,
      a: rgbaMatch[4] ? parseFloat(rgbaMatch[4]) : 1
    };
  }
  const hex = color.replace("#", "");
  if (hex.length === 3) {
    return {
      r: parseInt(hex[0] + hex[0], 16) / 255,
      g: parseInt(hex[1] + hex[1], 16) / 255,
      b: parseInt(hex[2] + hex[2], 16) / 255,
      a: 1
    };
  }
  if (hex.length === 6) {
    return {
      r: parseInt(hex.slice(0, 2), 16) / 255,
      g: parseInt(hex.slice(2, 4), 16) / 255,
      b: parseInt(hex.slice(4, 6), 16) / 255,
      a: 1
    };
  }
  return { r: 1, g: 0.89, b: 0.56, a: 0.5 };
}
function scaledToPdfPoints(scaled, page) {
  const pdfWidth = page.getWidth();
  const pdfHeight = page.getHeight();
  const xRatio = pdfWidth / scaled.width;
  const yRatio = pdfHeight / scaled.height;
  const x = scaled.x1 * xRatio;
  const width = (scaled.x2 - scaled.x1) * xRatio;
  const height = (scaled.y2 - scaled.y1) * yRatio;
  const y = pdfHeight - scaled.y1 * yRatio - height;
  return { x, y, width, height };
}
function dataUrlToBytes(dataUrl) {
  const base64 = dataUrl.split(",")[1];
  const byteString = atob(base64);
  const bytes = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    bytes[i] = byteString.charCodeAt(i);
  }
  const type = dataUrl.includes("image/png") ? "png" : "jpg";
  return { bytes, type };
}
function wrapText(text, font, fontSize, maxWidth) {
  if (!text || maxWidth <= 0) return [];
  const lines = [];
  const paragraphs = text.split(/\n/);
  for (const paragraph of paragraphs) {
    if (!paragraph.trim()) {
      lines.push("");
      continue;
    }
    const words = paragraph.split(/\s+/);
    let currentLine = "";
    for (const word of words) {
      const testLine = currentLine ? `${currentLine} ${word}` : word;
      const testWidth = font.widthOfTextAtSize(testLine, fontSize);
      if (testWidth <= maxWidth) {
        currentLine = testLine;
      } else {
        if (currentLine) {
          lines.push(currentLine);
          currentLine = "";
        }
        if (font.widthOfTextAtSize(word, fontSize) > maxWidth) {
          let remaining = word;
          while (remaining.length > 0) {
            let charCount = 1;
            while (charCount < remaining.length && font.widthOfTextAtSize(remaining.substring(0, charCount + 1), fontSize) <= maxWidth) {
              charCount++;
            }
            const chunk = remaining.substring(0, charCount);
            remaining = remaining.substring(charCount);
            if (remaining.length > 0) {
              lines.push(chunk);
            } else {
              currentLine = chunk;
            }
          }
        } else {
          currentLine = word;
        }
      }
    }
    if (currentLine) lines.push(currentLine);
  }
  return lines;
}
function groupByPage(highlights) {
  const map = /* @__PURE__ */ new Map();
  for (const h of highlights) {
    const pageNum = h.position.boundingRect.pageNumber;
    if (!map.has(pageNum)) map.set(pageNum, []);
    map.get(pageNum).push(h);
  }
  return map;
}
async function renderTextHighlight(page, highlight, options) {
  const colorStr = highlight.highlightColor || options.textHighlightColor || "rgba(255, 226, 143, 0.5)";
  const color = parseColor(colorStr);
  const highlightStyle = highlight.highlightStyle || "highlight";
  const rects = highlight.position.rects.length > 0 ? highlight.position.rects : [highlight.position.boundingRect];
  for (const rect of rects) {
    const { x, y, width, height } = scaledToPdfPoints(rect, page);
    if (highlightStyle === "highlight") {
      page.drawRectangle({
        x,
        y,
        width,
        height,
        color: rgb(color.r, color.g, color.b),
        opacity: color.a
      });
    } else if (highlightStyle === "underline") {
      const lineThickness = Math.max(1, height * 0.1);
      page.drawRectangle({
        x,
        y,
        width,
        height: lineThickness,
        color: rgb(color.r, color.g, color.b),
        opacity: color.a
      });
    } else if (highlightStyle === "strikethrough") {
      const lineThickness = Math.max(1, height * 0.1);
      const lineY = y + height / 2 - lineThickness / 2;
      page.drawRectangle({
        x,
        y: lineY,
        width,
        height: lineThickness,
        color: rgb(color.r, color.g, color.b),
        opacity: color.a
      });
    }
  }
}
async function renderAreaHighlight(page, highlight, options) {
  const colorStr = highlight.highlightColor || options.areaHighlightColor || "rgba(255, 226, 143, 0.5)";
  const color = parseColor(colorStr);
  const { x, y, width, height } = scaledToPdfPoints(
    highlight.position.boundingRect,
    page
  );
  page.drawRectangle({
    x,
    y,
    width,
    height,
    color: rgb(color.r, color.g, color.b),
    opacity: color.a
  });
}
async function renderFreetextHighlight(page, highlight, options, font) {
  const text = highlight.content?.text || "";
  const textColor = parseColor(
    highlight.color || options.defaultFreetextColor || "#333333"
  );
  const { x, y, width, height } = scaledToPdfPoints(
    highlight.position.boundingRect,
    page
  );
  const pdfHeight = page.getHeight();
  const yRatio = pdfHeight / highlight.position.boundingRect.height;
  const storedFontSize = parseInt(highlight.fontSize || "") || options.defaultFreetextFontSize || 14;
  const fontSize = storedFontSize * yRatio;
  console.log("Freetext export:", {
    storedFontSize,
    yRatio,
    fontSize,
    boxDimensions: { x, y, width, height },
    text: text.substring(0, 50)
  });
  const bgColorValue = highlight.backgroundColor || options.defaultFreetextBgColor || "#ffffc8";
  if (bgColorValue !== "transparent") {
    const bgColor = parseColor(bgColorValue);
    page.drawRectangle({
      x,
      y,
      width,
      height,
      color: rgb(bgColor.r, bgColor.g, bgColor.b),
      opacity: bgColor.a
    });
  }
  const padding = 4 * yRatio;
  const maxWidth = width - padding * 2;
  const lineHeight = fontSize * 1.3;
  if (maxWidth > 0 && text) {
    const lines = wrapText(text, font, fontSize, maxWidth);
    let currentY = y + height - fontSize - padding;
    for (const line of lines) {
      if (currentY < y + padding) break;
      if (line.trim()) {
        page.drawText(line, {
          x: x + padding,
          y: currentY,
          size: fontSize,
          font,
          color: rgb(textColor.r, textColor.g, textColor.b)
        });
      }
      currentY -= lineHeight;
    }
  }
}
function transformToRawCoordinates(page, x, y, width, height) {
  const rotation = page.getRotation().angle;
  const pageWidth = page.getWidth();
  const pageHeight = page.getHeight();
  if (rotation === 90) {
    return {
      x: y,
      y: pageWidth - x - width,
      width: height,
      height: width
    };
  } else if (rotation === 180) {
    return {
      x: pageWidth - x - width,
      y: pageHeight - y - height,
      width,
      height
    };
  } else if (rotation === 270) {
    return {
      x: pageHeight - y - height,
      y: x,
      width: height,
      height: width
    };
  }
  return { x, y, width, height };
}
async function renderImageHighlight(pdfDoc, page, highlight) {
  const imageDataUrl = highlight.content?.image;
  if (!imageDataUrl) return;
  try {
    const { bytes, type } = dataUrlToBytes(imageDataUrl);
    const image = type === "png" ? await pdfDoc.embedPng(bytes) : await pdfDoc.embedJpg(bytes);
    const visualCoords = scaledToPdfPoints(
      highlight.position.boundingRect,
      page
    );
    const rawCoords = transformToRawCoordinates(
      page,
      visualCoords.x,
      visualCoords.y,
      visualCoords.width,
      visualCoords.height
    );
    console.log("Image export:", {
      rotation: page.getRotation().angle,
      visualCoords,
      rawCoords
    });
    page.drawImage(image, {
      x: rawCoords.x,
      y: rawCoords.y,
      width: rawCoords.width,
      height: rawCoords.height
    });
  } catch (error) {
    console.error("Failed to embed image:", error);
  }
}
async function renderShapeHighlight(page, highlight) {
  const shapeType = highlight.content?.shape?.shapeType || highlight.shapeType || "rectangle";
  const strokeColorStr = highlight.content?.shape?.strokeColor || highlight.strokeColor || "#000000";
  const strokeWidth = highlight.content?.shape?.strokeWidth || highlight.strokeWidth || 2;
  const color = parseColor(strokeColorStr);
  const { x, y, width, height } = scaledToPdfPoints(
    highlight.position.boundingRect,
    page
  );
  switch (shapeType) {
    case "rectangle":
      page.drawRectangle({
        x,
        y,
        width,
        height,
        borderColor: rgb(color.r, color.g, color.b),
        borderWidth: strokeWidth,
        opacity: color.a
      });
      break;
    case "circle":
      page.drawEllipse({
        x: x + width / 2,
        y: y + height / 2,
        xScale: width / 2,
        yScale: height / 2,
        borderColor: rgb(color.r, color.g, color.b),
        borderWidth: strokeWidth,
        opacity: color.a
      });
      break;
    case "arrow": {
      const startPt = highlight.content?.shape?.startPoint;
      const endPt = highlight.content?.shape?.endPoint;
      const startX = startPt ? x + startPt.x * width : x;
      const startY = startPt ? y + (1 - startPt.y) * height : y + height / 2;
      const endX = endPt ? x + endPt.x * width : x + width;
      const endY = endPt ? y + (1 - endPt.y) * height : y + height / 2;
      page.drawLine({
        start: { x: startX, y: startY },
        end: { x: endX, y: endY },
        color: rgb(color.r, color.g, color.b),
        thickness: strokeWidth,
        opacity: color.a
      });
      const angle = Math.atan2(endY - startY, endX - startX);
      const arrowSize = Math.min(15, width * 0.2, height * 0.4);
      const arrowAngle = Math.PI / 6;
      page.drawLine({
        start: {
          x: endX - arrowSize * Math.cos(angle - arrowAngle),
          y: endY - arrowSize * Math.sin(angle - arrowAngle)
        },
        end: { x: endX, y: endY },
        color: rgb(color.r, color.g, color.b),
        thickness: strokeWidth,
        opacity: color.a
      });
      page.drawLine({
        start: {
          x: endX - arrowSize * Math.cos(angle + arrowAngle),
          y: endY - arrowSize * Math.sin(angle + arrowAngle)
        },
        end: { x: endX, y: endY },
        color: rgb(color.r, color.g, color.b),
        thickness: strokeWidth,
        opacity: color.a
      });
      break;
    }
  }
}
async function exportPdf(pdfSource, highlights, options = {}) {
  let pdfBytes;
  if (typeof pdfSource === "string") {
    const response = await fetch(pdfSource);
    pdfBytes = await response.arrayBuffer();
  } else {
    pdfBytes = pdfSource instanceof Uint8Array ? pdfSource.buffer.slice(
      pdfSource.byteOffset,
      pdfSource.byteOffset + pdfSource.byteLength
    ) : pdfSource;
  }
  const pdfDoc = await PDFDocument.load(pdfBytes);
  const pages = pdfDoc.getPages();
  const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
  const byPage = groupByPage(highlights);
  const totalPages = byPage.size;
  let currentPage = 0;
  for (const [pageNum, pageHighlights] of byPage) {
    const page = pages[pageNum - 1];
    if (!page) continue;
    for (const highlight of pageHighlights) {
      switch (highlight.type) {
        case "text":
          await renderTextHighlight(page, highlight, options);
          break;
        case "area":
          await renderAreaHighlight(page, highlight, options);
          break;
        case "freetext":
          await renderFreetextHighlight(page, highlight, options, font);
          break;
        case "image":
          await renderImageHighlight(pdfDoc, page, highlight);
          break;
        case "drawing":
          await renderImageHighlight(pdfDoc, page, highlight);
          break;
        case "shape":
          await renderShapeHighlight(page, highlight);
          break;
        default:
          await renderAreaHighlight(page, highlight, options);
      }
    }
    currentPage++;
    options.onProgress?.(currentPage, totalPages);
  }
  return pdfDoc.save();
}
export {
  exportPdf
};
